syntax = "proto2";

import "TSPMessages.proto";
package TSP;

message ArchiveInfo {
  optional uint64 identifier = 1;
  repeated .TSP.MessageInfo message_infos = 2;
  optional bool should_merge = 3;
}

message MessageInfo {
  required uint32 type = 1;
  repeated uint32 version = 2 [packed = true];
  required uint32 length = 3;
  repeated .TSP.FieldInfo field_infos = 4;
  repeated uint64 object_references = 5 [packed = true];
  repeated uint64 data_references = 6 [packed = true];
  optional uint32 base_message_index = 7;
  repeated uint32 diff_merge_version = 8 [packed = true];
  optional .TSP.FieldPath diff_field_path = 9;
  repeated .TSP.FieldPath fields_to_remove = 10;
  repeated uint32 diff_read_version = 11 [packed = true];
}

message FieldInfo {
  enum Type {
    Value = 0;
    ObjectReference = 1;
    DataReference = 2;
    Message = 3;
  }
  enum UnknownFieldRule {
    IgnoreAndPreserveUntilModified = 0;
    IgnoreAndPreserve = 1;
    MustUnderstand = 2;
    NotSupported = -1;
  }
  enum KnownFieldRule {
    None = 0;
    PreserveNewerValueUntilModified = 1;
    PreserveNewerValue = 2;
  }
  required .TSP.FieldPath path = 1;
  optional .TSP.FieldInfo.Type type = 2 [default = Value];
  optional .TSP.FieldInfo.UnknownFieldRule unknown_field_rule = 3 [default = IgnoreAndPreserveUntilModified];
  repeated uint64 object_references = 4 [packed = true];
  repeated uint64 data_references = 5 [packed = true];
  optional .TSP.FieldInfo.KnownFieldRule known_field_rule = 6 [default = None];
  repeated uint32 known_field_version = 7 [packed = true];
  optional string known_field_feature_identifier = 8;
}

message FieldPath {
  repeated uint32 path = 1 [packed = true];
}

message ComponentInfo {
  required uint64 identifier = 1;
  required string preferred_locator = 2;
  optional string locator = 3;
  repeated uint32 document_read_version = 4 [packed = true];
  repeated uint32 document_write_version = 5 [packed = true];
  repeated .TSP.ComponentExternalReference external_references = 6;
  repeated .TSP.ComponentDataReference data_references = 7;
  optional bool is_stored_outside_object_archive = 10 [default = false];
  repeated .TSP.ObjectUUIDMapEntry object_uuid_map_entries = 11;
  optional uint64 save_token = 12 [default = 0];
  repeated .TSP.FeatureInfo feature_infos = 13;
  repeated uint32 component_read_version = 14 [packed = true];
  repeated uint32 component_required_version = 15 [packed = true];
  optional uint32 compression_algorithm = 16;
  optional bool can_be_dropped = 17;
  repeated .TSP.ComponentExternalReference versioned_external_references = 18;
  optional bool is_wasteful = 19;
  repeated uint64 ambiguous_object_identifiers = 20 [packed = true];
  optional uint32 required_package_identifier = 21;
}

message ComponentExternalReference {
  required uint64 component_identifier = 1;
  optional uint64 object_identifier = 2;
  optional bool is_weak = 3;
}

message ComponentDataReference {
  required uint64 data_identifier = 1;
}

message ObjectUUIDMapEntry {
  required uint64 identifier = 1;
  required .TSP.UUID uuid = 2;
}

message FeatureInfo {
  required string identifier = 1;
  repeated uint32 read_version = 2 [packed = true];
  repeated uint32 write_version = 3 [packed = true];
}

message PackageMetadata {
  enum PackageType {
    Default = 0;
    Directory = 1;
    SingleFile = 2;
  }
  required uint64 last_object_identifier = 1;
  optional .TSP.DocumentRevision revision = 2;
  repeated .TSP.ComponentInfo components = 3;
  repeated .TSP.DataInfo datas = 4;
  repeated uint32 read_version = 5 [packed = true];
  repeated uint32 write_version = 6 [packed = true];
  repeated uint32 file_format_version = 7 [packed = true];
  optional uint64 save_token = 8 [default = 0];
  optional .TSP.PackageMetadata.PackageType preferred_package_type = 9 [default = Default];
  optional .TSP.Reference data_metadata_map = 10;
  repeated .TSP.ComponentInfo versioned_components = 11;
}

message DocumentRevision {
  optional int32 sequence_32 = 3 [default = 0];
  optional string identifier = 2;
  optional uint64 sequence_64 = 1 [default = 0];
}

message PasteboardMetadata {
  repeated uint32 version = 1 [packed = true];
  required string app_name = 2;
  repeated .TSP.DataInfo datas = 3;
  optional .TSP.UUID source_document_uuid = 5;
  optional .TSP.Reference data_metadata_map = 6;
  repeated uint32 read_version = 7 [packed = true];
}

message DataInfo {
  enum DownloadPriority {
    High = 0;
    Default = 1;
  }
  required uint64 identifier = 1;
  required bytes digest = 2;
  required string preferred_file_name = 3;
  optional string file_name = 4;
  optional string document_resource_locator = 5;
  optional bytes source_bookmark_data = 6;
  optional string remote_url = 7;
  optional bool can_download = 8 [default = false];
  optional .TSP.DataInfo.DownloadPriority download_priority = 9 [default = Default];
  optional .TSP.DataAttributes attributes = 10;
  optional .TSP.EncryptionInfo encryption_info = 11;
  optional bytes last_mismatched_digest = 12;
  optional string pasteboard_external_file_path = 99;
}

message DataMetadataMap {
  message DataMetadataMapEntry {
    required uint64 data_identifier = 1;
    required .TSP.Reference data_metadata = 2;
  }
  repeated .TSP.DataMetadataMap.DataMetadataMapEntry data_metadata_entries = 1;
}

message DataMetadata {
  optional .TSP.Color fallback_color = 1;
}

message EncryptionInfo {
  optional uint64 decoded_length = 1;
  optional uint64 preferred_block_size = 2;
  repeated .TSP.EncryptionBlockInfo block_infos = 3;
}

message EncryptionBlockInfo {
  required uint64 encoded_length = 1;
  optional uint64 decoded_length = 2;
}

message ViewStateMetadata {
  repeated uint32 version = 1 [packed = true];
  optional .TSP.UUID version_uuid = 4;
  required .TSP.ComponentInfo component = 3;
  repeated .TSP.ObjectUUIDMapEntry external_object_uuid_map_entries = 5;
  repeated uint32 read_version = 6 [packed = true];
}

message ObjectSerializationMetadata {
  repeated uint32 version = 1 [packed = true];
  optional .TSP.UUID source_document_uuid = 2;
  optional .TSP.UUID version_uuid = 3;
  required .TSP.ComponentInfo component = 4;
  repeated .TSP.DataInfo datas = 5;
  repeated .TSP.ObjectUUIDMapEntry external_object_uuid_map_entries = 6;
  optional .TSP.Reference data_metadata_map = 7;
  repeated uint32 read_version = 8 [packed = true];
}

message ObjectSerializationDirectory {
  message Entry {
    required string locator = 1;
    required uint64 offset = 2;
    required uint64 size = 3;
  }
  repeated .TSP.ObjectSerializationDirectory.Entry entries = 1;
}

message DataPropertiesEntryV1 {
  required bytes digest = 1;
  optional bool expects_matched_digest = 2;
  optional double creation_time_interval_since_1970 = 3;
  repeated uint32 creation_version = 4 [packed = true];
  optional bytes last_mismatched_digest = 5;
}

message DataPropertiesV1 {
  repeated .TSP.DataPropertiesEntryV1 properties = 1;
}

message DocumentMetadata {
  optional bool is_in_collaboration_mode = 1;
  optional .TSP.DataPropertiesV1 data_properties_v1 = 3;
}

message SupportMetadata {
  optional bool is_in_collaboration_mode = 1;
  repeated .TSP.DataReference acknowledged_server_data = 2;
}

